File: 2022/lib/day_01.ex
defmodule Day01 do
  def part_1(input) do
    input
    |> format
    |> Enum.max
  end

  def part_2(input) do
    input
    |> format
    |> Enum.sort
    |> Enum.take(-3)
    |> Enum.sum
  end

  defp format(input) do
    input
    |> String.split("\n")
    |> Enum.reduce([], &reduce_fn(&1, &2))
  end

  defp reduce_fn(x, []), do: [String.to_integer(x)]
  defp reduce_fn("", acc), do: [0 | acc]
  defp reduce_fn(x, acc) do
    [head | tail] = acc
    [head + String.to_integer(x) | tail]
  end
end

File: 2024/day_01.rb
# frozen_string_literal: true

require './advent_day'
require 'debug'

class Day01 < AdventDay
  LEFT = 0
  RIGHT = 1

  def part_1!
    left, right = 
      @data.reduce([[], []]) do |acc, current|
        l, r = current.split(' ').map(&:strip)
        [
          (acc[LEFT] + [l.to_i]).sort, 
          (acc[RIGHT] + [r.to_i]).sort
        ]
      end

    left.map.with_index { |_, index| (left[index] - right[index]).abs }.sum
  end

  def part_2!
    left, right = 
      @data.reduce([[], []]) do |acc, current|
        l, r = current.split(' ').map(&:strip)
        [
          (acc[LEFT] + [l.to_i]), (acc[RIGHT] + [r.to_i])
        ]
      end

    left.map { |l| l * right.count(l) }.sum
  end
end
